### PR TITLE
Turing Classes for Review

### PR DESCRIPTION
**Pull request by rodriguezjoanna**

### CODE REVIEW

**Review by graceKC04 — COMMENTED** [review]
# Code Review: RandomAccessSet

## Design

### Informative Identifiers:
All functions and variables are clearly named, making the code easy to follow. One improvement would be in `getIndexOf()`, which currently performs a linear search on elements. Since `indexMap` already stores element positions, using `indexMap.find(item)` could make this lookup more efficient. Additionally, consider adopting a consistent naming convention for member variables—such as adding a trailing underscore (`indexMap_`) or prefix (`mIndexMap`)—to distinguish them from local variables.

### Avoids “Magic Numbers”:
No arbitrary numbers are used in the code, so it meets best practices in this area.

### Intuitive and Functional Interface:
The class code has clear functionality. Everything seems useful and straightforward.

### Informative Comments:
The code is very well-commented, and I especially appreciate the detailed class description. The only comment that might be unnecessary is:
```cpp
`elements.push_back(item); // Store element in order`  
```
Since the purpose of `push_back` is already clear from the code itself, this comment could be omitted. Additionally, if you plan to use Doxygen for documentation in the future, adding `@brief`, `@param`, and `@return` annotations to function descriptions could be beneficial.

### Consistent Style:
Within `RandomAccessSet`, the style is very consistent. However, across the project, I noticed a few inconsistencies:
- Header guard naming conventions vary. Establishing a project-wide standard (e.g., `PROJECT_CLASSNAME_H`) would improve readability.
- Function naming conventions are mixed—some use camelCase, others use PascalCase. It would be helpful to standardize this. For instance, standard library functions use `.size()`, so keeping that makes sense, but `getIndexOf()` might benefit from following the same convention as other project functions.

### Visual Appearance:
The code is easy to read and well-structured. Thoughtful comments make the functionality very clear.

## Implementation

### No Warnings During Compilation:
The code compiles cleanly without errors, which is great.

### Proper Use of Namespaces:
`RandomAccessSet` is not enclosed in a `cse` namespace. Make sure code is enclosed in `namespace cse{}`. 

### Adheres to the C++ Standard:
The implementation avoids undefined behavior and correctly handles exceptions where necessary.

### Appropriate Modularity:
Functions are well-structured and modular, avoiding unnecessary duplication.

### Use of Modern C++ Constructs:
The code is well-written and already utilizes modern C++ features effectively. I don’t see any immediate areas for improvement.

### Use of Standard Library Algorithms:
There are no areas where I see standard library algorithms significantly improving clarity or efficiency. The implementation aligns well with best practices.

### Pointer Handling:
Pointers are not used in this implementation, so it meets best practices.

### Const Correctness:
Functions that do not modify the object are correctly marked `const`, following best practices.

## Testing and Error Handling

### Unit Testing:
Most important cases are covered. One additional test that could be useful is checking object construction. Even though defaults are currently used, a test case would be helpful in case the constructor logic changes in the future.

### Edge Cases:
A few additional edge cases could strengthen test coverage:
- A test where `getIndexOf()` is called with an invalid input to confirm that it properly throws and catches an exception.
- Ensuring `get()` correctly handles incorrect inputs by verifying that an exception is both thrown and handled within the test case.
- Testing `remove()` for specific edge cases:
  - Removing the last element in `elements`.
  - Removing an element at index 0, ensuring it correctly swaps and updates `indexMap`.

### Assertions for Debugging:
The code is well-structured, but adding assertions could help catch inconsistencies between `indexMap` and `elements`. For example, in `add()`, you could assert that an element does not already exist before inserting it:
```cpp
assert(indexMap.find(item) == indexMap.end() && "Attempting to add a duplicate element!");
```

Similarly, after remove(), an assertion could check whether indexMap and elements remain synchronized:
```cpp
assert(indexMap.size() == elements.size() && "Mismatch between elements vector and indexMap after removal!");
```
These assertions wouldn’t affect production code but could be useful in debug mode to catch subtle bugs early.

### Error Handling
- The code correctly accounts for invalid inputs. In further development, it may be useful to provide clear feedback to the user when an error occurs. You could consider:
  - Returning a boolean or error code from functions instead of silently failing.
  - Adding a logging mechanism to provide meaningful error messages.
  - Expanding exception handling to provide more context about failures.

### Overall Summary
- The code is well-structured, efficient, and easy to read. A few improvements—like creating consistency across class files—could be useful in increasing readability.
- Class test cases should also probably be moved within the overarching tests folder.

Let me know if you have any additional questions or clarifications!

**Comment by Devon-FoxElster** [comment]
# Review for RandomAccessSet

## Design
### Informative identifiers
The name of the functions are descriptive enough, as well as most of the variables used. I would recommend renaming some of the iterator variables to something more descriptive of what they are iterating over, as currently their names do not give any indication of their purpose.
### Avoids “magic numbers”
I did not see the use of any magic numbers so this is good to go.
### Intuitive and functional interface
Overall I think the interface is good, although get() as a function feels like it would be retrieving a specific value rather than accessing an index. Possibly renaming get() to at()?
### Informative comments
I particularly like the main class comment, as I think it outlines the purpose and capabilities of the class very well. There were a few comments that can probably be deleted, such as the one Grace mentioned as well as:
`template <typename T> // Template class allows RAS to work with any data type`
I'm also a big fan of Doxygen comments for member functions, so using @param, @brief, @return, and @tparam is recommended in the .cpp file.
### Consistent style
The style is camelCase, and is very consistent throughout this class. I would also recommend an identifier for the member variables, with a _ after the variable names or an m before.
### Visual appearance
Clean and organized, overall very nice.

## Implementation
### No warnings during compilation
Compiling with the test cases went through with no errors, but directly compiling the RandomAccessSet.cpp file did pull up some warnings. 
- The algorithm std::find was used in getIndexOf but the library <algorithm> isn't included in either of the class files. 
- Similar error with the out_of_range exception, including <stdexcept> takes care of this in the header or cpp file.

### Proper uses of namespaces
 I did not see the use of namespace cse for the class, so I recommend adding that to the class.
 
### Adheres to the C++ standard
All functions seem to handle exceptions well and limit undefined behavior.
### Appropriate Modularity
The code seems separated well, I did see one opportunity to use contains() within getIndexOf. You could use contains() to confirm that the element is within the Set, otherwise throwing your error. Grace's suggestion of using indexMap in that function rather than std::find is also a great suggestion.
### Uses modern C++ constructs
Good use of modern C++ practices
### Uses standard library algorithms
The use of std::find and std::swap were appropriate in their contexts
### Pointer Handling
No pointers in use, so this one is good as well.
### Const Correctness
Const correctness seems to be consistent across the class.

## Testing and Error Checking
### Unit testing
The constructor could use a test case to confirm it is empty upon initialization, just for backwards-compatibility purposes if changes were made in the future.
I would also move the size() test case earlier in the test file, just because that function is frequently used within the other test cases that are listed before it.
### Edge cases
Edge cases are covered within the class functions, but testing to ensure that the class functions cover them correctly should be added in the case of getIndexOf() and get()
### Assertions
It seems that exceptions are primarily used in the code rather than asserts, but they cover most potential errors possible in the functions.
### Error handling
Errors are handled well within the implementation.


### Other suggestions
I also recommend that the tests be moved into the class-wide test folder, as Professor Ofria mentioned setting up a system that would run all of the tests together to keep consistency. Other than that I think the test structure you and your group members set up works very well.

I also think the templated functions would work better being stored within the .h file, as when I compiled it I did get notes about the .cpp functions overriding the ones present within the .h folder.
`RandomAccessSet.cpp:9:6: note: redefinition of ‘bool RandomAccessSet<T>::add(const T&)’
    9 | bool RandomAccessSet<T>::add(const T& item) {`

**Comment by merryymary** [comment]
**Design**
• **Informative identifiers:** Names of functions like Append and ToString are very straight forward. Variable mSegments clearly denotes it is a member variable as well. One suggestion is to rename the str_fun_t to be more identifiable as to its purpose.
• **Avoids “magic numbers”:** No magic numbers were used.
• **Intuitive and functional interface:** It seems pretty intuitive of how to use Append and ToString. The only thing that was difficult to understand at first was the overloaded operator+, how there was one for string and one for a function string. But further looking into the code it makes more sense, so maybe more comments around their differences/use cases is unnecessary.
• **Informative comments:** There are good comments through out each function. I did notice some redundant ones like on line 119: result += segValue; // append the string
• Comments like this can clutter the more important understanding ones, so maybe re read through and make sure comments are on the why of code or explaining complex issues.
• **Consistent style:** Consistent style of PascalCase and using m as a predecessor for member variables.
• **Visual appearance:** Easy to read and consistent spacing.

**Implementation**
• **No warnings during compilation:** Produced one warning regarding line 129, how there is no return statement for a non void function ToString(). 
• **Proper uses of namespaces:** There is no namespace cse, so make sure to add that around the header and cpp file, but after the #include statements.
• **Adheres to the C++ standard:** Uses of std::function appear correct and up to standard.
• **Appropriate Modularity:** I couldn't find any redundant loops or room for helper functions, as each function is pretty clearly stated and coded.
• **Uses modern C++ constructs:** You use std::function, std::visit, and std::variant to shorten the code.
• **Uses standard library algorithms:** There seems to only be a loop for traversing each segment, so these are necessary. Otherwise you use std::visit to access each vector segment.
• **Pointer handling:** No raw pointers.
• **const correctness:** ToString and Append are correctly const marked.

**Testing and Error Checking**
• **Unit testing:** I couldn't see a test specifically for DyanamicString. Make sure to add some test to test each function and constructor.
• **Edge cases:** Make sure to add edge cases for making sure errors are properly caught. For example, an empty string being added. An empty string as a parameter for ToString. A null function being appended. 
• **Assertions:** I couldn't find any assertions in the code. One could be place in the constructor to make sure str_fun_t is not null. Also an assert could be lace in the ToString() function to make sure mSegments isn't empty before you try to access it.
• **Error handling:** There could be a runtime error when using std::visit function. I would recommend adding a try catch surround that in case it fails it won't fully break the program.

File: Group-08/Dev/DynamicString/DynamicString.cpp [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+                result += segValue; // append the string
+            }
+            // If the segment is a function, call it
+            else
+            {
+                // Call function to get the updated text and append it
+                result += segValue();
+            }
+        }, segment);
+    }
```
Comment (merryymary) [inline Group-08/Dev/DynamicString/DynamicString.cpp:128]: Make sure to return the result here

File: Group-08/Dev/DynamicString/DynamicString.cpp [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        std::visit(
+            // A lambda that gets a reference to our result string
+            // segValue represents the value held in std::variant (segment)
+            [&result](auto &&segValue) {
+            // Determines the type of the segment
+            using segType = std::decay_t<decltype(segValue)>;
+            // Checks if the segment is a std::string
+            if constexpr (std::is_same_v<segType, std::string>)
+            {
+                result += segValue; // append the string
```
Comment (merryymary) [inline Group-08/Dev/DynamicString/DynamicString.cpp:119]: Maybe this comment is unneeded. It seems pretty clear the segValue is being appended onto the result string. It may take away from other informative comments

**Comment by Muelleralex1021** [comment]
### Review for AnnotatedWrapper
**Design**
Informative identifiers
all names of variables, functions and types are clear and have a consistent format. everything here looks good.

Avoids “magic numbers”
I did not see the use of any magic numbers so this is good.

Intuitive and functional interface
Class is very straight-forward and only allows for necessary functionality.
Function names are clear to how they are intended to be used. 

Informative comments
Comments from function to function are all informative. 
I would suggest adding an overall class comment that describes potentially the use of the class in an example or how it is intended to be used. 
Maybe consider adding a small comment for the annotations member variable in the header file. 

Consistent style
The style is camelCase, and is very consistent throughout this class.

Visual appearance
everything looks great.

**Implementation**
No warnings during compilation
Compiling with the test cases went through with no errors.
Makefile in test folder does not work. 
The TEST_NAMES values are for AuditedPointer and not AnnotatedWrapper!

Does not have function inside namespace cse.

Adheres to the C++ standard
All functions seem to handle exceptions well and limit undefined behavior.

Appropriate Modularity
looks good. not really necessary as functions are all very short. 

Uses modern C++ constructs
Good use of modern C++ practices

Uses standard library algorithms
all good.

Pointer Handling
No pointers in use, so this one is good as well.

Const Correctness
Const correctness seems to be consistent across the class.

**Testing and Error Checking**
Unit testing
tests are all present for each function besides list which just prints anyway but could be included.

Edge cases
could check for number of keys in annotations after using clearAnnotations. 

Assertions
Require is used instead of ASSERT.

Error handling
doesn't seem that there is any.

**Comment by ephraimbennett** [comment]
# Review for Audited Pointer
## Design
### Informative Identifiers.
Names of variables, functions are mostly clear. It seems to be consistent naming across the files. It doesn't appear that too many local variables are necessary anyways. Perhaps the parameter for copy and move operations, ie. "other" could be renamed more explicitly but it seems to be fine. 
### Avoids “magic numbers”.
There are no magic numbers present in the templated AuditedPointer.h file, yet there exist many usages of them throughout both test files and main.cpp. This could be fine, but it is recommended to not use them. 
### Intuitive and functional interface.
The class exposes the necessary functionality, with a global tracker, standalone MakeAudited function as well. Member variables are private. The use of operator overloading is not necessarily confusing, but could be more clear what is being done for copy constructor and copy assignment operator overloads vs move constructor and move assignment operator overloads. **It seems that the class is lacking a .New() member function (or alternatively an overload of the new keyword).** 
### Informative comments.
While there are some comments throughout the class which provide insight, brief explanations on the logic behind the copy overrides and move overrides would have been helpful for understanding. Additionally, consider using doxygen comments to provide clarity.
### Consistent style. 
Style choices are used in a consistent manner. Variables use camelCase, while functions and class names use PascalCase, and member variables are prefixed with m.
### Visual appearance.
The code is easy to scan through and visually well structured, with a limit on line widths. 
## Implementation
### No warnings during compilation.
The code compiles with the -Wall and -Wextra flags without producing warnings.
### Proper uses of namespaces.
The code is **not** wrapped in namespace cse {}. The global namespace is **not** pulluted by using namespace std;
### Adheres to the C++ standard.
The code adheres to the C++ standard and avoids undefined behavior on different compilers and platforms. 
### Appropriate Modularity.
Common procedures are grouped into useful helper functions such as MakeAudited. Helper structs such as LeakChcker are utilized. There is very little to no code duplication.
### Uses modern C++ constructs. 
There are not any instances where the usage of range-based for loops, or std::optional could be utilitized. Templates are utilized correctly, and the move constructors and assignment operators are implemented as well.
### Uses standard library algorithms.
The code does not require the usage of loops and does need need to implement these types of standard library algorithms.
### Pointer handling.
Memory management is done correctly. Raw pointers are deleted when necesarry and destructors are utilized. Since this is a wrapper class for raw pointers with it's own unique functionality, clearly smart pointers are not needed here.
### const correctness. 
Member functions are marked const that do not change the state of the class. Parameters that are not changed are marked const, the ones that are (such as in the move operators) are not marked const.
## Testing and Error Checking
### Unit testing. 
Unit tests are present for for public emmebr functions and trigger cases throughout the classes code.

**Comment by owenhaiar** [comment]
# Code Review: AuditedPointer 

## Design

### Informative identifiers:
The identifiers are clear and to the point, good job

### Avoids “magic numbers”:
No magic numbers were needed in source file. The numbers in the test files is fine in my opinion, the numbers are arbitrary and its clear how they’re being used

### Intuitive and functional interface:
The interface is great, almost everything is handled using overloaded operators. Consider overloading the new and delete keywords instead of using .Delete(). 

### Informative comments:
Use doxygen style documentation instead of one line comments above functions. Also add documentation at the top of the header.

### Consistent style / Visual appearance:
The style is consistent, but the large amount of whitespace hurts readability. The constructor, for example, could be “Aptr(TYPE* ptr = nullptr) : mPtr(ptr) { if (mPtr) ++globalCounter; }” instead of 7 lines. Also consider using a different style for the global variable, like ALL_CAPS


## Implementation

### No warnings during compilation:
No warnings, but you do have some unreachable code in MakeAudited

### Proper uses of namespaces:
Use namespace cse for your code

### Adheres to the C++ standard:
No undefined behavior (apart from dereferencing a nullptr in non-debug but a normal pointer acts the same). Looks good

### Appropriate Modularity:
The code is really simple, so nothing to be done here

### Uses modern C++ constructs:
Good job with templating

### Uses standard library algorithms:
Nothing to do here, there are no algorithms in the class

### Pointer handling:
No memory management issues

### Const correctness:
Functions are marked const where appropriate 

## Testing and Error Checking

### Unit testing:
Consider adding test cases for the arrow operator, and for data types other than ints. You should probably also have test cases for memory leaks. Also, I had to edit the makefile to get AuditedPointerNonDebugTest to run. 

### Edge cases:
Add test cases for exiting when trying to dereference a nullptr in debug mode.

### Assertions:
The debug mode pointer has assertions in good spots

### Error handling:
No obvious changes here, problems in this class should abort instead of throwing an exception

---

Functionally the code looks good, the main issues are the documentation and the limited test cases.

**Comment by miabodenhorn** [comment]
### **TagManager Review**

#### **Design**
- **Informative identifiers:** Names are clear and easy to understand.  
- **Avoids magic numbers:** No hard-coded numbers in the code, which is great. Everything is stored in maps and sets, so no concerns there.  
- **Intuitive and functional interface:** The class is easy to use, and the functions are well named.  
- **Informative comments:** Comments straightforward and useful.  
- **Consistent style:** The code is formatted consistently.  
- **Visual appearance:** Everything is spaced well, and the indentation is clean.  

#### **Implementation**
- **No warnings during compilation:** Compiled with no warnings.  
- **Proper use of namespaces:** The code doesn’t use a namespace, which is fine for this project, but adding one (like `cse`) might help avoid name conflicts.  
- **Adheres to the C++ standard:** The code follows best practices.  
- **Appropriate modularity:** Functions are structured well, and everything is in the right place. No unnecessary duplication.  
- **Uses modern C++ constructs/standard library:** Uses `std::unordered_map` and `std::unordered_set`, which are great choices for fast lookups.  
- **Pointer handling:** No raw pointers, which is great!  
- **Const correctness:** Functions that return data but don’t modify it could be marked `const`, like `GetTags()` and `GetTaskTags()`. Also, the function parameters that don’t change should be marked as `const`, like `bool HasTag(const string& task, const string& tag) const`.  

#### **Testing and Error Checking**
- **Unit testing:** The provided tests cover most cases, including:  
  - Adding and removing tags.  
  - Checking if a task has a tag.  
  - Clearing tags.  
  - Edge cases like removing non-existent tags.  
- **Edge cases:** The tests seem to cover common edge cases, but testing for empty tasks or tags would be good.  
- **Assertions:** I didn’t see any assertions in the code, but adding some in functions like `RemoveTag()` (to check if the task exists before modifying it) could help catch bugs earlier.  
- **Error handling:** The code handles missing tags and tasks well by returning empty sets. However, adding error messages or exceptions in extreme cases (like passing empty strings as entries or tags) might be useful
### **Final Thoughts**
The code is nicely structured, and the class does what it's supposed to! Nice job.

**Comment by AmeryJohnson** [comment]
# Turing - DynamicString Review
**Design**
**Informative Identifiers:** Informative besides the type alias str_fun_t 
**Avoids Magic Numbers:** Yes
**Intuitive and functional Interface:** I think it checks all the boxes for this category it is pretty intuitive.
**Informative Comments:** Comments are informative 
**Consistent Style:** I believe so nothing popped out that was inconsistent.
**Visual Appearance:** Has good spacing and comments are consistent and put in good places.

**Implementation**
**No warnings during compilation:** There is a warning on the ToString function since it does not return a string when it should.
**Proper uses of Namespace:** Needs cse namespace 
**Adheres to the C++ Standard:** Yes
**Appropriate Modularity:** Yes there isn't really redundant code
**Uses modern C++ constructs:** Good uses of the standard namespace tools like function, variant, and visit
**Uses standard library algorithms:** Yes std::visit seems to be put to good use here
**Pointer handling:** No pointers
**Const correctness:** Everything marked const seems to be correct.

**Testing and Error Checking** 
**Unit Testing:** I could not find any unit tests in the folder
**Edge Cases:** Should check for empty strings and other edge cases that could arise when testing
**Assertions:** No assertions in the code. Could add some for a lot of edge cases.
**Error Handling:** I do not see much error handling within the code which could break the program if error arise which will most likely happen. Checking if something is a non string or empty strings etc.

**Comment by Bcrims** [comment]
### TagManager Review
Design
Informative identifiers: The names used are clear and self-explanatory, making it easy to understand what each variable and function represents
Avoids magic numbers: There are no hard-coded numbers; the code relies on standard containers for its operations
Intuitive and functional interface: The class offers a straightforward interface with well-named functions, though marking read-only functions as const could further improve usability.
Informative comments: The comments are useful in explaining the purpose of functions
Consistent style: The formatting, naming, and indentation are consistent throughout the code.
Visual appearance: The code is neatly organized with clear spacing and clean indentation.

Implementation
No warnings during compilation: The code compiles with no warnings
Proper use of namespaces: Currently, the code is in the global namespace instead of the CSE namespace that was recommended.
Adheres to the C++ standard: The code follows modern C++ practices and standards.
Appropriate modularity: The functionality is well partitioned into separate header and source files, with no redundant code.
Uses modern C++ constructs/standard library: The use of std::unordered_map and std::unordered_set is appropriate and takes advantage of modern C++ libraries.
Pointer handling: There is no use of raw pointers, which reduces the risk of memory issues.
Const correctness: Accessor functions like GetTags, GetTaskTags, and HasTag should be marked as const since they do not modify the object’s state

Testing and Error Checking
Unit testing: The main.cpp example demonstrates basic functionality and the tests cover most cases
Edge cases: The code handles some edge cases such as removing a tag from a task that doesn’t exist, additional tests for empty strings or unexpected input could be beneficial.
Assertions: No assertions are present, adding them in functions could catch errors during development
Error handling: The code handles missing data by returning empty sets

The TagManager class is well-designed and implements its intended functionality effectively.

**Review by graceKC04 — COMMENTED** [review]
# Code Review: RandomAccessSet

## Design

### Informative Identifiers:
All functions and variables are clearly named, making the code easy to follow. One improvement would be in `getIndexOf()`, which currently performs a linear search on elements. Since `indexMap` already stores element positions, using `indexMap.find(item)` could make this lookup more efficient.

Additionally, consider adopting a consistent naming convention for member variables—such as adding a trailing underscore (`indexMap_`) or prefix (`m_indexMap`)—to distinguish them from local variables.

### Avoids "Magic Numbers":
No arbitrary numbers are used in the code, so it meets best practices in this area.

### Intuitive and Functional Interface:
The class code has clear functionality. Everything seems useful and straightforward.

### Informative Comments:
The code is very well-commented, and I especially appreciate the detailed class description. The only comment that might be unnecessary is:

```cpp
elements.push_back(item); // Store element in order  
```
Since the purpose of `push_back` is already clear from the code itself, this comment could be omitted.


Additionally, if you plan to use Doxygen for documentation in the future, adding `@brief`, `@param`, and `@return` annotations to function descriptions could be beneficial.

## Consistent Style

Within `RandomAccessSet`, the style is very consistent. However, across the project, I noticed a few inconsistencies:

- **Header guard naming conventions vary.** Establishing a project-wide standard (e.g., `PROJECT_CLASSNAME_H`) would improve readability.
- **Function naming conventions are mixed**—some use `camelCase`, others use `PascalCase`. It would be helpful to standardize this. For instance, standard library functions use `.size()`, so keeping that makes sense, but `getIndexOf()` might benefit from following the same convention as other project functions.

## Visual Appearance

The code is easy to read and well-structured. Thoughtful comments make the functionality very clear.

## Implementation

### No Warnings During Compilation

The code compiles cleanly without errors, which is great.

### Proper Use of Namespaces

`RandomAccessSet` is not enclosed in a `cse` namespace.

### Adheres to the C++ Standard

The implementation avoids undefined behavior and correctly handles exceptions where necessary.

### Appropriate Modularity

Functions are well-structured and modular, avoiding unnecessary duplication.

### Use of Modern C++ Constructs

The code is well-written and already utilizes modern C++ features effectively. I don’t see any immediate areas for improvement.

### Use of Standard Library Algorithms

There are no areas where I see standard library algorithms significantly improving clarity or efficiency. The implementation aligns well with best practices.

### Pointer Handling

Pointers are not used in this implementation, so it meets best practices.

### Const Correctness

Functions that do not modify the object are correctly marked `const`, following best practices.

## Testing and Error Handling

### Unit Testing

Most important cases are covered. One additional test that could be useful is checking object construction. Even though defaults are currently used, a test case would be helpful in case the constructor logic changes in the future.

### Edge Cases

A few additional edge cases could strengthen test coverage:

- A test where `getIndexOf()` is called with an invalid input to confirm that it properly throws and catches an exception.
- Ensuring `get()` correctly handles incorrect inputs by verifying that an exception is both thrown and handled within the test case.
- Testing `remove()` for specific edge cases:
  - Removing the last element in `elements`.
  - Removing an element at index `0`, ensuring it correctly swaps and updates `indexMap`.

### Assertions for Debugging

The code is well-structured, but adding assertions could help catch inconsistencies between `indexMap` and `elements`. For example, in `add()`, you could assert that an element does not already exist before inserting it:

```cpp
assert(indexMap.find(item) == indexMap.end() && "Attempting to add a duplicate element!"); 
```
Similarly, after remove(), an assertion could check whether indexMap and elements remain synchronized:
```cpp 
assert(indexMap.size() == elements.size() && "Mismatch between elements vector and indexMap after removal!");
 ```

These assertions wouldn’t affect production code but could be useful in debug mode to catch subtle bugs early.

### Error Handling
- The code correctly accounts for invalid inputs. In further development it may be useful to provide clear feedback to the user when an error occurs.  You could consider:
   - Returning a boolean or error code from functions instead of silently failing.
   - Adding a logging mechanism to provide meaningful error messages.
   - Expanding exception handling to provide more context about failures.
---
Overall, the code is well-structured, efficient, and easy to read. A few improvements—like creating consistency across class files could be useful in increasing readability. Class test cases should also probably be moved within the overarching tests folder. Let me know if you have any additional questions or clarifications!
